File: widgetssdk/src/main/java/com/glia/widgets/core/engagement/domain/ShouldShowMediaEngagementViewUseCase.java
package com.glia.widgets.core.engagement.domain;

import com.glia.widgets.engagement.domain.EngagementTypeUseCase;
import com.glia.widgets.engagement.domain.IsQueueingOrLiveEngagementUseCase;

/**
 * @hide
 */
public class ShouldShowMediaEngagementViewUseCase {
    private final IsQueueingOrLiveEngagementUseCase isQueueingOrLiveEngagementUseCase;
    private final EngagementTypeUseCase engagementTypeUseCase;

    public ShouldShowMediaEngagementViewUseCase(
        IsQueueingOrLiveEngagementUseCase isQueueingOrLiveEngagementUseCase,
        EngagementTypeUseCase engagementTypeUseCase) {
        this.isQueueingOrLiveEngagementUseCase = isQueueingOrLiveEngagementUseCase;
        this.engagementTypeUseCase = engagementTypeUseCase;
    }

    public boolean execute(boolean isUpgradeToCall) {
        return hasNoQueueingAndEngagementOngoing() ||
            hasMediaQueueingOngoing() ||
            hasOngoingMediaEngagement() ||
            isUpgradeToCall;
    }

    private boolean hasNoQueueingAndEngagementOngoing() {
        return !isQueueingOrLiveEngagementUseCase.invoke();
    }

    private boolean hasMediaQueueingOngoing() {
        return isQueueingOrLiveEngagementUseCase.isQueueingForMedia();
    }

    private boolean hasOngoingMediaEngagement() {
        return engagementTypeUseCase.isMediaEngagement();
    }

}

File: widgetssdk/src/main/java/com/glia/widgets/internal/engagement/domain/ShouldShowMediaEngagementViewUseCase.kt
package com.glia.widgets.internal.engagement.domain

import com.glia.widgets.engagement.domain.EngagementTypeUseCase
import com.glia.widgets.engagement.domain.IsQueueingOrLiveEngagementUseCase

internal class ShouldShowMediaEngagementViewUseCase(
    private val isQueueingOrLiveEngagementUseCase: IsQueueingOrLiveEngagementUseCase,
    private val engagementTypeUseCase: EngagementTypeUseCase
) {
    fun execute(isUpgradeToCall: Boolean): Boolean {
        return hasNoQueueingAndEngagementOngoing() ||
            hasMediaQueueingOngoing() ||
            hasOngoingMediaEngagement() ||
            isUpgradeToCall
    }

    private fun hasNoQueueingAndEngagementOngoing(): Boolean {
        return !isQueueingOrLiveEngagementUseCase.invoke()
    }

    private fun hasMediaQueueingOngoing(): Boolean {
        return isQueueingOrLiveEngagementUseCase.isQueueingForMedia
    }

    private fun hasOngoingMediaEngagement(): Boolean {
        return engagementTypeUseCase.isMediaEngagement
    }
}
